### DISCUSSION TITLE
obtaining abstract syntax tree

### DISCUSSION THREAD

**Discussion by nhasabni** [MikePopoloski/slang#444] [Q&A] [answered]
Hi,

I was using `SyntaxTree::parseFile` and `SyntaxNode::toString` functions to print a syntax tree, but I saw that this ends up printing a concrete syntax tree with actual variable names. Is there API or a way to print an abstract syntax tree? Or I would need to traverse the tree and print it myself? Thanks.

**Comment by MikePopoloski** [comment, marked as answer]
I think of the SyntaxTree / SyntaxNode as the "parse tree" or equivalently "concrete syntax tree" as you mention. The *abstract* syntax tree is built out of symbols, expressions, statements, etc by the Compilation object. That "tree" is serializable to JSON via the ASTSerializer class but it's done in a kind of ad-hoc / not very generalizable fashion so if you want something other than JSON you need to walk the tree yourself (the ASTVisitor class can be of help here).

**Reply by nhasabni** [reply]
Okay, thanks.

**Reply by nhasabni** [reply]
Hi Mike,

I was able to write a simple tree traversal to print AST for a SyntaxNode. It works nicely in most cases, but for some reason, it is printing a space, when I am printing a token.

E.g., for expression ```wb_wr | rd_fifo_becoming_empty```, the AST is printed is 
```(ConditionalPattern(BinaryOrExpression(IdentifierName(wb_wr))(IdentifierName( rd_fifo_becoming_empty))))```

Notice the space before `rd_fifo_becoming_empty` in AST. It looks to me that the token includes space as well. Am I doing something incorrect?

Here is my print function:
```
 std::string Print(const slang::SyntaxNode& node) const {
    std::ostringstream stream;
    if (slang::toString(node.kind).length() > 0)
      stream << "(" << node.kind;

    auto print_token_for_syntax_kinds = [](slang::SyntaxKind kind) {
      // Print tokens for identifiers and literals
      return ((slang::NameSyntax::isKind(kind)) ||
              (slang::LiteralExpressionSyntax::isKind(kind)));
    };

    for (uint32_t i = 0; i < node.getChildCount(); i++) {
      if (node.childNode(i)) {
        stream << Print(*node.childNode(i));
      } else if (node.childToken(i)) {
        if (print_token_for_syntax_kinds(node.kind)) {
          slang::SyntaxPrinter syntax_printer;
          syntax_printer.print(node.childToken(i));
          stream << "(" << syntax_printer.str() << ")";
        }
      }
    }

    if (slang::toString(node.kind).length() > 0)
      stream << ")";
    return stream.str();
 }
```

Thanks.

**Reply by MikePopoloski** [reply]
The SyntaxPrinter class will, by default, print trivia when printing a token (trivia is stuff like whitespace and comments that come before the token). You can disable that with setIncludeTrivia(false).

**Reply by nhasabni** [reply]
Ah, I should have checked the doc. My bad. That worked. Thanks.
